File: R/Descriptive Analytics/ConductKMeansClustering.R
# Load packages
library(dplyr)
library(ggplot2)
library(mclust)

# Declare function
ConductKMeansClustering = function(dataframe,
                                   list_of_variables_to_create_clusters,
                                   standardize_variables = TRUE,
                                   number_of_clusters = NULL,
                                   column_name_for_clusters = "kmeans_cluster",
                                   show_cluster_modeling_results = TRUE,
                                   random_seed = 412) {
  # Set seed, if specified
  if (!is.null(random_seed)) {
    set.seed(random_seed)
  }
  
  # Select variables specified, and keep complete cases only
  dataframe_clust = dataframe %>% select(
    one_of(list_of_variables_to_create_clusters)
  ) %>% na.omit
  
  # Standardize variables
  if (standardize_variables) {
    dataframe_clust = scale(dataframe_clust)
    dataframe_clust = as.data.frame(dataframe_clust)
  }
  
  # If number_of_clusters not specified, use mclust to identify the most likely model and number of clusters
  if (is.null(number_of_clusters)) {
    cluster_model = Mclust(dataframe_clust)
    number_of_clusters = cluster_model$G
    if (show_cluster_modeling_results) {
      print(cluster_model$BIC)
    }
  }
  
  # Conduct k-means clustering
  cluster_results = kmeans(dataframe_clust, number_of_clusters)
  
  # Bind clusters to original dataframe
  df_clusters = as.data.frame(cluster_results$cluster)
  dataframe = bind_cols(
    dataframe,
    df_clusters
  ) 
  names(dataframe)[names(dataframe) == "cluster_results$cluster"] = column_name_for_clusters
  
  # Convert to cluster to factor type
  dataframe[[column_name_for_clusters]] = as.factor(dataframe[[column_name_for_clusters]])
  
  # Import box-whisker function from SnippetsForStatistics
  source("https://raw.githubusercontent.com/onenonlykpro/SnippetsForStatistics/master/R/Visualizations/PlotBoxWhiskerByGroup.R")
  
  # Visualize variables within each cluster using boxplots
  for (variable in list_of_variables_to_create_clusters) {
    PlotBoxWhiskerByGroup(dataframe = dataframe,
                          quantitative_variable = variable,
                          grouping_variable_1 = column_name_for_clusters)
  }
  
  # Return dataframe
  return(dataframe)
}

File: Python/Descriptive Analytics/ConductKMeansClustering.py
# Load packages
import pandas as pd
import random
from sklearn.cluster import KMeans
from sklearn.preprocessing import StandardScaler
from yellowbrick.cluster import KElbowVisualizer

# Declare function
def ConductKMeansClustering(dataframe,
                            list_variables_to_form_clusters = None,
                            number_of_clusters = None,
                            standardize_variables = True,
                            random_seed = 412,
                            column_name_for_clusters = 'K-Means Cluster'):
    # Select variables specified
    if list_variables_to_form_clusters == None:
        list_variables_to_form_clusters = dataframe.columns
        df_for_clustering = dataframe.copy()
    else:
        df_for_clustering = dataframe[list_variables_to_form_clusters].copy()

    # Keep complete cases only
    df_for_clustering = df_for_clustering.dropna()

    # Standardize variables
    if standardize_variables:
        df_transformed = StandardScaler().fit_transform(df_for_clustering)
        df_transformed = pd.DataFrame(df_transformed)
        df_transformed.index = df_for_clustering.index
        df_for_clustering = df_transformed.copy()
        df_for_clustering.columns = list_variables_to_form_clusters

    # If number of clusters not specified, use ___ to find "best" number
    if number_of_clusters == None:
        model = KMeans()
        visualizer = KElbowVisualizer(
            model,
            k=(2,30),
            timings=True
        )
        visualizer.fit(df_for_clustering)
        visualizer.show()
        number_of_clusters = visualizer.elbow_value_
        
    # Conduct k-means clusters
    model = KMeans(
        n_clusters=number_of_clusters, 
        random_state=random_seed
    )
    model = model.fit(df_for_clustering)

    # Join clusters to original dataset
    df_for_clustering[column_name_for_clusters] = model.labels_

    # Convert cluster to categorical data type
    df_for_clustering[column_name_for_clusters] = df_for_clustering[column_name_for_clusters].apply(str)

    # If requested, show box plots of each cluster for each variable selected

    # Join clusters to original dataset
    df_for_clustering = df_for_clustering [[column_name_for_clusters]]
    dataframe = dataframe.merge(
        df_for_clustering,
        how='left',
        left_index=True,
        right_index=True
    )

    # Return updated dataset
    return(dataframe)
